File: C++/Leetcode 1472.cpp
#include <stdio.h>
#include <bits/stdc++.h>
using namespace std;

class BrowserHistory {
    stack<string> Backward;
    stack<string> Forward;
public:
    BrowserHistory(string homepage) {
        Backward.push(homepage);
    }
    
    void visit(string url) {
        Backward.push(url);
        while (Forward.size())
        {
            Forward.pop();
        }
    }
    
    string back(int steps) {
        while (steps && Backward.size())
        {
            Forward.push(Backward.top());
            steps--;
            Backward.pop();
        }

        if (!Backward.size())
        {
            Backward.push(Forward.top());
            Forward.pop();
        }
        return Backward.top();
    }
    
    string forward(int steps) {
        while (steps && Forward.size())
        {
            steps--;
            Backward.push(Forward.top());
            Forward.pop();
        }

        return Backward.top();
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */

File: Python/Leetcode 1472.py
class BrowserHistory:
    Forward = []
    Backward = []

    def __init__(self, temp_string: str):
        # print(self.Backward)

        while len(self.Backward):
            self.Backward.pop()

        while len(self.Forward):
            self.Forward.pop()

        # print(temp_string)
        self.Backward.append(temp_string)
        # print(len(self.Backward))
        # print(self.Backward[0])

    def visit(self, url: str) -> None:
        self.Backward.append(url)

        while len(self.Forward):
            self.Forward.pop()

    def back(self, steps: int) -> str:
        while steps and len(self.Backward):
            steps = steps - 1
            self.Forward.append(self.Backward[len(self.Backward) - 1]) 
            self.Backward.pop()

        if len(self.Backward) == 0:
            self.Backward.append(self.Forward[len(self.Forward) - 1])
            self.Forward.pop()
        
        return self.Backward[len(self.Backward) - 1]

    def forward(self, steps: int) -> str:
        while steps and len(self.Forward):
            steps = steps - 1
            self.Backward.append(self.Forward[len(self.Forward) - 1])
            self.Forward.pop()
        
        return self.Backward[len(self.Backward) - 1]
        
        

# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)